File: packages/acp-py/maia_acp/stream.py
"""ACP stream utilities — parse raw SSE into typed ACP events."""

from __future__ import annotations

import json
import logging
from typing import Any, Generator

import httpx

from maia_acp.types import ACPEvent

logger = logging.getLogger(__name__)


def parse_sse_line(line: str) -> ACPEvent | None:
    """Parse a single SSE line into an ACPEvent.

    Returns None for non-data lines, keep-alive, or [DONE].
    Gracefully wraps non-ACP JSON payloads as activity events.
    """
    trimmed = line.strip()
    if not trimmed or not trimmed.startswith("data: "):
        return None

    data_str = trimmed[6:]
    if data_str == "[DONE]":
        return None

    try:
        parsed = json.loads(data_str)
    except json.JSONDecodeError:
        return None

    if not isinstance(parsed, dict):
        return None

    # Native ACP event
    if parsed.get("acp_version"):
        return ACPEvent.model_validate(parsed)

    # Non-ACP event — wrap intelligently
    return _wrap_non_acp(parsed)


def _wrap_non_acp(data: dict[str, Any]) -> ACPEvent:
    """Wrap a non-ACP JSON payload into an ACP envelope.

    This enables Theatre to render events from ANY existing SSE stream,
    even if the backend doesn't know about ACP.
    """
    agent_id = (
        data.get("agent_id")
        or data.get("agent")
        or data.get("sender")
        or data.get("name")
        or "unknown"
    )
    agent_uri = agent_id if agent_id.startswith("agent://") else f"agent://{agent_id}"

    content = (
        data.get("content")
        or data.get("message")
        or data.get("text")
        or data.get("output")
        or json.dumps(data, default=str)
    )

    event_type = "event"
    if data.get("content") or data.get("message") or data.get("text"):
        event_type = "message"

    if event_type == "message":
        payload = {
            "from": agent_uri,
            "to": "agent://user",
            "intent": "propose",
            "content": str(content),
        }
    else:
        payload = {
            "agent_id": agent_uri,
            "activity": "tool_calling",
            "detail": str(content),
        }

    return ACPEvent(
        run_id=str(data.get("run_id", data.get("thread_id", "unknown"))),
        agent_id=agent_uri,
        event_type=event_type,  # type: ignore[arg-type]
        payload=payload,
    )


def stream_events(lines: Generator[str, None, None]) -> Generator[ACPEvent, None, None]:
    """Convert a line generator into an ACPEvent generator."""
    for line in lines:
        event = parse_sse_line(line)
        if event is not None:
            yield event


def connect_sse(
    url: str,
    *,
    headers: dict[str, str] | None = None,
    timeout: float = 120.0,
) -> Generator[ACPEvent, None, None]:
    """Connect to an SSE endpoint and yield ACP events.

    Works with both ACP-native and non-ACP SSE streams.

    Usage:
        for event in connect_sse("http://localhost:8000/acp/events"):
            if event.event_type == "message":
                msg = event.as_message()
                print(f"{msg.from_agent}: {msg.content}")
    """
    request_headers = {
        "Accept": "text/event-stream",
        **(headers or {}),
    }

    with httpx.Client(timeout=httpx.Timeout(timeout)) as client:
        with client.stream("GET", url, headers=request_headers) as response:
            response.raise_for_status()
            for line in response.iter_lines():
                event = parse_sse_line(line)
                if event is not None:
                    yield event

File: packages/acp-js/src/stream.ts
// ─── ACP Stream Utilities ───────────────────────────────────────────────────
// Parse raw SSE streams into typed ACP events.
// Works with any SSE source — not just ACP-native servers.

import type { ACPEvent } from "./types";

/**
 * Parse a raw SSE line into an ACP event.
 * Returns null if the line is not a valid ACP event.
 * Gracefully handles non-ACP SSE data by wrapping it as an activity event.
 */
export function parseSSELine(line: string): ACPEvent | null {
  const trimmed = line.trim();
  if (!trimmed || !trimmed.startsWith("data: ")) return null;

  const dataStr = trimmed.slice(6);
  if (dataStr === "[DONE]") return null;

  try {
    const parsed = JSON.parse(dataStr);

    // Native ACP event — pass through
    if (parsed.acp_version) {
      return parsed as ACPEvent;
    }

    // Non-ACP event — try to wrap it intelligently
    return wrapNonACPEvent(parsed);
  } catch {
    return null;
  }
}

/**
 * Wrap a non-ACP JSON event into an ACP envelope.
 * This is what makes Theatre work with ANY existing SSE stream.
 */
function wrapNonACPEvent(data: Record<string, unknown>): ACPEvent {
  // Try to detect common patterns from popular frameworks
  const agentId =
    (data.agent_id as string) ??
    (data.agent as string) ??
    (data.sender as string) ??
    (data.name as string) ??
    "agent://unknown";

  const agentUri = agentId.startsWith("agent://")
    ? agentId
    : `agent://${agentId}`;

  // Detect event type from common fields
  let eventType: ACPEvent["event_type"] = "event";
  if (data.content || data.message || data.text) {
    eventType = "message";
  } else if (data.tool_call || data.tool || data.function_call) {
    eventType = "event";
  }

  const content =
    (data.content as string) ??
    (data.message as string) ??
    (data.text as string) ??
    (data.output as string) ??
    JSON.stringify(data);

  return {
    acp_version: "1.0",
    run_id: (data.run_id as string) ?? (data.thread_id as string) ?? "unknown",
    agent_id: agentUri,
    event_type: eventType,
    timestamp: (data.timestamp as string) ?? new Date().toISOString(),
    payload:
      eventType === "message"
        ? {
            from: agentUri,
            to: "agent://user",
            intent: "propose" as const,
            content,
          }
        : {
            agent_id: agentUri,
            activity: "tool_calling" as const,
            detail: content,
          },
  };
}

/**
 * Async generator that reads a ReadableStream (e.g. fetch response body)
 * and yields ACP events.
 */
export async function* streamToACPEvents(
  stream: ReadableStream<Uint8Array>,
): AsyncGenerator<ACPEvent> {
  const reader = stream.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";

      for (const line of lines) {
        const event = parseSSELine(line);
        if (event) yield event;
      }
    }

    // Process remaining buffer
    if (buffer.trim()) {
      const event = parseSSELine(buffer);
      if (event) yield event;
    }
  } finally {
    reader.releaseLock();
  }
}

/**
 * Connect to an SSE endpoint and yield ACP events.
 * Works with both ACP-native and non-ACP SSE streams.
 */
export async function* connectToSSE(
  url: string,
  options?: RequestInit,
): AsyncGenerator<ACPEvent> {
  const response = await fetch(url, {
    ...options,
    headers: {
      Accept: "text/event-stream",
      ...options?.headers,
    },
  });

  if (!response.ok) {
    throw new Error(`ACP stream connection failed: ${response.status}`);
  }

  if (!response.body) {
    throw new Error("No response body");
  }

  yield* streamToACPEvents(response.body);
}
